src/groupeddataframe/splitapplycombine.jl

```diff
@@ -853,13 +853,58 @@ function fillfirst!(condf, outcol::AbstractVector, incol::AbstractVector,
     outcol
 end
 
-# Use reducedim_init to get a vector of the right type,
-# but trick it into using the expected length
-groupreduce_init(op, condf, incol, gd) =
-    Base.reducedim_init(identity, op, view(incol, 1:length(gd)), 2)
+# Use a strategy similar to reducedim_init from Base to get the vector of the right type,
+function groupreduce_init(op, condf, adjust,
+                          incol::AbstractVector{U}, gd::GroupedDataFrame) where U
+    T = Base.promote_union(U)
+
+    if op isa typeof(Base.add_sum)
+        initf = zero
+    elseif op isa typeof(Base.mul_prod)
+        initf = one
+    else
+        throw(ErrorException("Unrecognized op $op"))
+    end
+
+    Tnm = nonmissingtype(T)
+    if Tnm !== Any && isconcretetype(Tnm) && applicable(initf, Tnm)
+        tmpv = initf(Tnm)
+        initv = op(tmpv, tmpv)
+        x = adjust isa Nothing ? initv : adjust(initv, 1)
+        if condf isa typeof(!ismissing)
+            V = typeof(x)
+        else
+            V = U >: Missing ? Union{typeof(x), Missing} : typeof(x)
+        end
+    else
+        idx = findfirst(!ismissing, incol)
+        # here the definition of x is simpler as we do V = Any anyway
+        if isnothing(idx)
+            if condf isa typeof(!ismissing)
+                throw(ArgumentError("Aggregating over a column containing " *
+                                    "only missing values that does not have a " *
+                                    "concrete union element type is not allowed"))
```

Comment:
Shouldn't we check `applicable(initf, T)`? As you noted, `zero(Real)` is defined so we could fill a `Vector{Any}` or `Vector{Real}` with that. Maybe even do that in all cases before looking for the first non-missing value as it's faster.

Reply:
OK - we can fall back here as you propose. However, we have to do `findfirst`, as otherwise we have the following problem:
```
julia> x = Real[UInt8(250), UInt8(250)]
2-element Array{Real,1}:
 0xfa
 0xfa

julia> sum(x)
0x00000000000001f4

julia> Base.add_sum(Base.add_sum(zero(Real), x[1]), x[2])
500
```
and we would be inconsistent with Base.

Reply:
> Maybe even do that in all cases before looking for the first non-missing value as it's faster.

In the end you are probably right and I will change to that. But the reason is different. We are not sure which values we will encounter in subgroups. I will propose yet a different implemenation

Reply:
Pondering on this even more I came to the conclusion that we should use yet a different strategy because of:
```
julia> sum(["a"])
"a"
```
I will try to propose something efficient for this (in general using `fillfirst!` I think is what we should do).

If the behaviour of `sum` is not intended then we should use a different strategy so please comment how you see this. Thank you!.

Reply:
I just implemented what Base does

Reply:
OK. It's too bad that we have to make things so complex to match what Base does. But in the end you didn't use `fillfirst!`, right? Wouldn't it be simpler for us?

Reply:
`fillfirst!` is used for `max` and `min` like operations as there it is safe. We could use `fillfirst!` for `add` and `prod` like operations but then the result of `fillfirst!` would have to be changed to `zero` or `one` in order to get a neutral element. This would mean that the "hard cases" at the end (which match Base behavior) would throw an error, so we would have an inconsistency between `combine(gdf, :x=>sum)` and `combine(gdf, :x=>(x->sum(x))` unfortunately. Also what I do now is faster than using `fillfirst!` (though the gain is probably non-noticeable).

Reply:
Ah yes this `sum(["a"])` thing is really absurd, but I don't see how Base could handle this differently.

One case where performance is going to be suboptimal is for small unions like `Union{Int8,Int32}` or `Union{Nothing,Int}`. It probably doesn't matter a lot, but in the future we could imagine adding a fast path using `fillfirst!` for `Number`, where we can assume `zero` and `one` are defined. Though even in these cases it would be faster only if the number of groups is relatively small I guess.

Reply:
Right, but I would not think that `Union{Int8,Int32}` or `Union{Nothing,Int}` are type specifications we should put a priority on - agreed?

Reply:
Right.